synapse/util/caches/__init__.py

```diff
@@ -195,8 +195,10 @@ def register_cache(
     )
 }
 
+T = TypeVar("T", Optional[str], str)
 
-def intern_string(string: Optional[str]) -> Optional[str]:
+
+def intern_string(string: T) -> T:
     """Takes a (potentially) unicode string and interns it if it's ascii"""
```

Comment:
Is the comment about the string being ascii relevant now? Ditto for the try/except.

I can't see any mention of either at https://docs.python.org/3/library/sys.html?highlight=sys%20intern#sys.intern . I'd guess this was some PY2 compat stuff we can just ditch?

The history seems to be
-  https://github.com/matrix-org/synapse/pull/2155
- https://github.com/matrix-org/synapse/pull/3247
- https://github.com/matrix-org/synapse/pull/7519